scripts/cleanup_redirects.rb

```diff
@@ -0,0 +1,193 @@
+  manual = JSON.parse(File.read('./_data/manual_redirects.json'))
+  generated.merge(manual)
+end
+
+# Gets a collection of redirects represented by the given markdown file.
+def parse_redirect(redirect_path)
+  raw = File.read(redirect_path).strip
+  # A redirect with content after the frontmatter is not "simple".
+  return unless raw.start_with?('---') && raw.end_with?('---')
+
+  frontmatter = YAML.load(raw)
+
+  return unless frontmatter.keys.sort == SIMPLE_KEYS
+
+  # Since the json plugin excludes things from the sitemap, we need to make sure
+  # the simple ones do, as well.
+  return if frontmatter['sitemap']
+
+  redirect_from = frontmatter['redirect_from']
+  redirect_from = [redirect_from] unless redirect_from.is_a?(Array)
+  redirect_to = frontmatter['redirect_to']
+  redirect_to = [redirect_to] unless redirect_to.is_a?(Array)
+
+  redirect_from.product(redirect_to).to_h
+end
+
+def determine_blockers(paths)
+  scripted = load_scripted_redirects
+  paths.flat_map do |redirect_path|
+    redirects = parse_redirect(redirect_path)
+    categorized = categorize_redirects(redirects, scripted)
+    (
+      handle_absent(categorized[:absent], scripted, redirect_path) +
+      handle_conflicts(categorized[:conflicts], scripted, redirect_path)
+    )
+  end
+end
+
+def check_validity(conflicts)
+  return conflicts unless @options[:check_validity]
+
+  conflicts.map do |conflict|
+    to = conflict[:md]
+    url = /^http/.match(to) ? to : File.join('https://ada.gov', to)
+
+    uri = URI.parse(url.gsub(' ', '%20'))
+    request = Net::HTTP.new(uri.host, uri.port)
+    request.use_ssl = true
+    begin
+      response = request.request_head(uri.path)
+      code = response.code
+      puts "Validated #{url} as #{code}"
+    rescue
+      code = -1
+    end
+
+    conflict.merge({status: code})
+  end
+end
+
+def handle_absent(redirects, scripted, path)
+  return [] if redirects.empty?
+
+  absent = redirects.map do |from, to|
+    { reason: 'absent', file: path, md: to, json: scripted[from] }
+  end
+
+  check_validity(absent)
+end
+
+def equal_ignoring_htm?(a, b)
+  a_extless = a.gsub(/\.html?$/, '')
+  b_extless = b.gsub(/\.html?$/, '')
+  return a_extless == b_extless
+end
+
+def handle_conflicts(redirects, scripted, path)
+  return [] if redirects.empty?
+
+  conflicts = redirects.map do |from, to|
+    { reason: 'conflict', file: path, md: to, json: scripted[from] }
+  end
+
+  conflicts = conflicts.select do |conflict|
+    next false if conflict[:json].downcase == conflict[:md].downcase
+
+    next false if equal_ignoring_htm?(conflict[:json], conflict[:md])
+
+    true
+  end
+
+  check_validity(conflicts)
+end
+
+def categorize_redirects(redirects, scripted)
+  present, absent = redirects.partition do |from, _|
+    scripted.key?(from)
+  end
+
+  matches, conflicts = present.partition do |from, to|
+    scripted[from] == to
+  end
+
+  {absent: absent.to_h, matches: matches.to_h, conflicts: conflicts.to_h}
+end
+
+def main
```

Comment:
I'm having a hard time working through this file. I think it would be easier for me if these functions were chunked into separate files and imported into the script file containing the main function. 

I think it would also be easier for me to understand how the individual functions map onto the 4 cases you outline at the start of the file. Not sure if it's appropriate to keep those cases in mind while understanding the code though.